### PR TITLE
Datetime enc fix dayOfWeek crash with Monday 00:00 

### LINKED ISSUE #578: Date Encoder crashes on encoding Mon 00:00
In the C++ code, if I understand correctly, right now the scalar encoder clips all the inputs to [min,max] and does not implement a periodic encoder. In lines 152-153 there is the following "clipping":
```
      input = std::max(input, parameters.minimum);
      input = std::min(input, parameters.maximum);
```
which only makes sure that input<=maximum. If input<=min, then input does not change.

This confuses me because I do not understand how the hotgym example is working, since there are certain times that need to be encoded that are out of bounds. An example is any time on a Monday before noon which causes dayOfWeek<=0 where [min,max]=[0,7].
I would be happy for any clarification on my misunderstanding, because clearly I am missing something.

Thank you!

### PR DESCRIPTION
**Pull request by breznak**
Fixes #578 
Part of reviewing DateTime #471

### CODE REVIEW

File: py/htm/encoders/date.py [hunk 1]
```diff
@@ -21,7 +21,7 @@
 import numpy
 import math
 
-from htm.encoders.scalar_encoder import ScalarEncoder, ScalarEncoderParameters
+from htm.bindings.encoders import ScalarEncoder, ScalarEncoderParameters
```
Comment (breznak) [inline py/htm/encoders/date.py]: we should use the bindings version (ie no pyplot dependency), moved scalar_encoder to examples.

File: py/htm/examples/scalar_encoder.py [hunk 2]
```diff
@@ -95,7 +95,9 @@
     else:
         n_samples = (enc.parameters.maximum - enc.parameters.minimum) / enc.parameters.resolution
         oversample = 2 # Use more samples than needed to avoid aliasing & artifacts.
-        n_samples  = int(round( oversample * n_samples ))
+        n_samples  = int(round( oversample * n_samples )) #TODO should this be done already in the c++?
```
Comment (breznak) [inline py/htm/examples/scalar_encoder.py]: do we have the concept of "category encoder" in c++? If so, the modification should be moved there

File: src/htm/encoders/ScalarEncoder.cpp [hunk 3]
```diff
@@ -159,7 +159,7 @@ void ScalarEncoder::encode(Real64 input, SDR &output)
         << "Input to category encoder must be an unsigned integer!";
     }
     NTA_CHECK(input >= parameters.minimum && input <= parameters.maximum)
-        << "Input must be within range [minimum, maximum]!";
+        << "Input must be within range [minimum, maximum]! " << input << " vs [" << parameters.minimum << " , " << parameters.maximum << " ]";
```
Comment (breznak) [inline src/htm/encoders/ScalarEncoder.cpp]: debuging `datetime` input from DateTime encoder

**Comment by breznak** [comment]
@shahardror  you can test with this branch, can you please add the test-case that triggers the issue? 
Also, what does `datetime.strptime("23/07/19 8:00", "%d/%m/%y %H:%M")` return? What does a negative value mean? 

The current output/err is: 

> ERR: CHECK FAILED: "input >= parameters.minimum && input <= parameters.maximum" Input must be within range [minimum, maximum]! -0.166667 vs [0 , 7 ] [/mnt/store/devel/HTM/htm-community/nupic.cpp/src/htm/encoders/ScalarEncoder.cpp line 161]
Traceback (most recent call last):
File "dateSizeCheck.py", line 14, in
Bits = Encoder.encode(date)
File "/mnt/store/devel/HTM/pyenv3/lib/python3.6/site-packages/htm.core-2.0.0-py3.6-linux-x86_64.egg/htm/encoders/date.py", line 260, in encode
sdrs.append( self.dayOfWeekEncoder.encode(dayOfWeek) )
RuntimeError: CHECK FAILED: "input >= parameters.minimum && input <= parameters.maximum" Input must be within range [minimum, maximum]! -0.166667 vs [0 , 7 ]

**Comment by shahardror** [comment]
> What does a negative value mean?

the dayOfWeek variable translates every time in a week to a number between [0,7] where Monday 00:00 is mapped to 0 and Sunday 00:00 to 6. But the dayOfWeek Encoder adds the following bias:
` dayOfWeek -= .5 # Round towards noon, not midnight`
which causes all the inputs on Mondays before 12:00 to be mapped to a negative value. 

> can you please add the test-case that triggers the issue?

I think I added it

File: py/htm/encoders/date.py (hunk 1 above)
Comment (ctrl-z-9000-times) [reply py/htm/encoders/date.py]: the scalar_encoder only imports `pyplot` if you try to use it.

File: py/htm/examples/scalar_encoder.py (hunk 2 above)
Comment (ctrl-z-9000-times) [reply py/htm/examples/scalar_encoder.py]: `n_samples` controls the number of inputs which it encodes to make the graphic.  It determines the which inputs are given to the encoder.

**Comment by breznak** [comment]
> the dayOfWeek variable translates every time in a week to a number between [0,7] where Monday 00:00 is mapped to 0 and Sunday 00:00 to 6. 

do you mean the encoder does the transformation to [0,7]?

> # Any Monday morning (before noon) # DEBUG
# d = datetime.strptime("22/07/19 8:00", "%d/%m/%y %H:%M") # DEBUG
bits = enc.encode(d)

but i meant what values `d` can be, and what would <0 mean? 

> But the dayOfWeek Encoder adds the following bias:
dayOfWeek -= .5 # Round towards noon, not midnight

great! this is the pitfall. 
- Should we remove the bias? But that would change semantics compared to Numenta's encoder.
  - actually, what does "Round towards noon, not midnight" mean? That days flip on noon, not midnight? I guess that's so unexpected we don't want that?!  
- we cannot `clipInput`-it, as that would .. -> we also could do that. 
- [ ] can you please add test cases for "Mon 00:00, Mon 23:59" have same repr, and "Tue 00:01" has different. And also try the same around noon, if we flipped around that.

File: py/htm/encoders/date.py (hunk 1 above)
Comment (breznak) [reply py/htm/encoders/date.py]: true. Is it still ok to use the bindings version, or is there any advantage to using the py wrapper?

**Comment by ctrl-z-9000-times** [comment]
> dayOfWeek -= .5 # Round towards noon, not midnight

I think the fix is to delete that line.  Since this encoder is not a category encoder, it does not round the number to the nearest integer.  Instead this is a periodic encoder, so shifting the input by `.5` does literally nothing to the information contained in the resulting SDR.

**Comment by shahardror** [comment]
> * But that would change semantics compared to Numenta's encoder.

I see that you previously discussed a similar topic in #462 . Perhaps it does impact the information because it is not implemented as a periodic encoder?

**Comment by breznak** [comment]
> Perhaps it does impact the information because it is not implemented as a periodic encoder? 

days & time are implemented as periodic enc, so removing should be fine (removing the bias seems better anyway) 

```
self.dayOfWeekEncoder = None
if dayOfWeek != 0:
p = ScalarEncoderParameters()
# Value is day of week (floating point)
# Radius is 1 day
p.minimum  = 0
p.maximum  = 7
p.periodic = True
```

File: py/htm/encoders/date.py (hunk 1 above)
Comment (ctrl-z-9000-times) [reply py/htm/encoders/date.py]: There is no advantage or disadvantage

**Comment by breznak** [comment]
@shahardror will you be able to finish this PR?

**Comment by breznak** [comment]
Merged the change (removal of bias) from @Alexsbuchanan , 

>  Instead this is a periodic encoder, so shifting the input by .5 does literally nothing to the information contained in the resulting SDR. 

this does not seem to be true, as tests break. I think it's fine to just adapt the expected bits in tests to match the current state.

File: py/htm/encoders/date.py [hunk 4]
```diff
@@ -249,12 +249,15 @@ def encode(self, inp, output=None):
...
+      assert(dayOfYear >= 0)
       # dayOfYear -= self.seasonEncoder.parameters.radius / 2. # Round towards the middle of the season.
       sdrs.append( self.seasonEncoder.encode(dayOfYear) )
 
     if self.dayOfWeekEncoder is not None:
-      dayOfWeek = timetuple.tm_wday + (timeOfDay) / 24.0
-      dayOfWeek -= .5 # Round towards noon, not midnight
+      hrs_ = float(timeOfDay) / 24.0 # add hours as decimal value in extension to day  
+      hrs_ = hrs_ * 0.5 # normalize hours to [0, 0.5], instead of [0, 1.0], because > 0.5 would cause rounding up to the next day
+      dayOfWeek = timetuple.tm_wday + hrs_
```
Comment (breznak) [inline py/htm/encoders/date.py]: fixed the hours, trick was to project hours only to 0..0,5, not 0..1.0, otherwise rounding skips to next day.

File: py/tests/encoders/date_test.py [hunk 5]
```diff
@@ -77,19 +77,26 @@ def testDayOfWeek(self):
...
+    # Monday = 0 (for python datetime.datetime.timetuple())
     dayOfWeekExpected = [0,0,0,1,0,0,0] #Thu
 
     expected = dayOfWeekExpected
     self.assertEqual(bits.size, 7)
     self.assertEqual(expected, bits.dense.tolist())
 
+    # check a day is encoded consistently during all of its hours (00:00-23:59):
+    dMorn = datetime.datetime(2010, 11, 4, 00, 00) #strptime("04/11/2010 0:00", "%d/%m/%y %H:%M")
+    dEve  = datetime.datetime(2010, 11, 4, 23, 59) #strptime("04/11/2010 23:59", "%d/%m/%y %H:%M")
```
Comment (breznak) [inline py/tests/encoders/date_test.py]: added test for edge cases of the day

File: py/htm/encoders/date.py (hunk 4 above)
Comment (ctrl-z-9000-times) [reply py/htm/encoders/date.py]: There is a flaw in this solution.  At midnight the representation looses its semantic similarity with the previous day. 

```python
import datetime
import numpy as np
from htm.encoders.date import DateEncoder
from htm import SDR, Metrics
enc = DateEncoder(
            season=0,
            dayOfWeek=300,
            weekend=0,
            holiday=0,
            timeOfDay=0,
            customDays=0,
            holidays=((12, 25),))
sdr = SDR( enc.dimensions )
test_period = 24 * 3
metrics = Metrics( sdr, test_period )
now = datetime.datetime.now()
inc = datetime.timedelta(hours=1)
ovlp = []
for i in range( test_period ):
    enc.encode(now, sdr)
    now += inc
    ovlp.append( metrics.overlap.overlap )

print( metrics )

import matplotlib.pyplot as plt
plt.plot( np.arange(len(ovlp)), ovlp )
plt.xlabel("Hours after now.")
plt.ylabel("Semantic Similarity")
plt.show()
```

Comment (ctrl-z-9000-times) [reply py/htm/encoders/date.py]:
```
>>> print( metrics )
```
SDR( 2100 )
    Sparsity Min/Mean/Std/Max 0.142857 / 0.142857 / 4.08069e-06 / 0.142857
    Activation Frequency Min/Mean/Std/Max 0 / 0.142856 / 0.143628 / 0.333333
    Entropy 0.743834
    Overlap Min/Mean/Std/Max 0.476667 / 0.958028 / 0.101115 / 0.98


Figure_1

Comment (breznak) [reply py/htm/encoders/date.py]: good catch and thanks for the nice test-case! :+1: 

> here is a flaw in this solution. At midnight the representation looses its semantic similarity with the previous day.

yep, that's the price for fixing the "bias" with that solution. 
Maybe the clearest option would be to let the encoder do just what it says: `dayOfWeek` is only for days (=no hours added), and `timeOfDay` fills the latter functionality. 
Btw, on this note, I've observed that in NAB optimizations, "timeOfDay" is useless if "dayOfWeek" is used with enough bits (and that is due to the `+hrs_` part).

File: bindings/py/tests/encoders/scalar_encoder_test.py [hunk 6]
```diff
@@ -23,8 +23,8 @@
 import pytest
 import time
 
-from htm.bindings.encoders import ScalarEncoder, ScalarEncoderParameters
-from htm.bindings.sdr import SDR, Metrics
+from htm.encoders.scalar_encoder import ScalarEncoder, ScalarEncoderParameters
+from htm import SDR, Metrics
```
Comment (breznak) [inline bindings/py/tests/encoders/scalar_encoder_test.py]: nit, why this version compared to the "pure" bindings? I assume some plot,debug functionality?

File: py/htm/encoders/date.py [hunk 7]
```diff
@@ -230,7 +230,7 @@ def encode(self, inp, output=None):
       output = SDR(self.dimensions)
     else:
       assert( isinstance(output, SDR) )
-      assert( output.dimensions == self.dimensions )
+      assert( all(x == y for x, y in zip( output.dimensions, self.dimensions )))
```
Comment (breznak) [inline py/htm/encoders/date.py]: that one was easir to read, why this?

File: py/htm/encoders/date.py (hunk 4 above)
Comment (breznak) [reply py/htm/encoders/date.py]: ok, I see the fix now. Not sure why I thought the `*0.5` is needed.

File: py/tests/encoders/date_test.py [hunk 8]
```diff
@@ -51,7 +52,7 @@ def testDateEncoder(self):
 
     # Week is MTWTFSS contrary to localtime documentation, Monday = 0 (for
     # python datetime.datetime.timetuple()
-    dayOfWeekExpected = [0,0,0,1,0,0,0]
+    dayOfWeekExpected = [0,0,0,0,1,0,0]
```
Comment (breznak) [inline py/tests/encoders/date_test.py]: We must review the correctness of the tests. The day should be Thrs, according to the comment above, the expected is now Fri, why? I think that was the reason I added the "normalization" to [0, 1/2].

Comment (breznak) [reply py/tests/encoders/date_test.py]: You've mentioned this is due to "small bitness" of the demo test, so that has to be mentioned in the comments. Preferably even changed so the encoding fits the logic that "Thursday afternoon" is still Thrs.

**Comment by breznak** [comment]
An afterthought on the similarity of "days", should we revert back to the "days are compared to noon", and just fix the case for Mon morning to "underflow" correctly to Sunday, wrt. similarity? Then representations "switch" and midnight (day-1:noon - day:noon)/2, which is more natural)

File: py/htm/encoders/date.py [hunk 9]
```diff
@@ -249,12 +250,18 @@ def encode(self, inp, output=None):
...
 
     if self.dayOfWeekEncoder is not None:
-      dayOfWeek = timetuple.tm_wday + (timeOfDay) / 24.0
-      dayOfWeek -= .5 # Round towards noon, not midnight
+      hrs_ = float(timeOfDay) / 24.0 # add hours as decimal value in extension to day  
+      dayOfWeek = timetuple.tm_wday + hrs_
+      dayOfWeek -= .5 # Round towards noon, not midnight, this means similarity of representations changes at midnights, not noon.
+      # handle underflow: on Mon before noon -> move to Sun
+      if dayOfWeek < 0:
+        dayOfWeek += 7
```
Comment (breznak) [inline py/htm/encoders/date.py]: A came with another fix, that reverts the `bias`, so days are compared against noon. 
- this is desirable so that dates are "similar" during day-time [morning, eve], and change representation at midnight.
- compatible behavior with Numenta's

The fix was to handle the underflow by +7 (aka comparing to prev Sunday)

File: py/tests/encoders/date_test.py [hunk 10]
```diff
@@ -77,19 +78,45 @@ def testDayOfWeek(self):
...
-    # d = datetime.datetime(2010, 11, 1, 8, 55) # DEBUG
-    # Any Monday morning (before noon) # DEBUG
-    # d = datetime.strptime("22/07/19 8:00", "%d/%m/%y %H:%M") # DEBUG
     bits = enc.encode(d)
 
-    # Week is MTWTFSS contrary to localtime documentation, Monday = 0 (for
-    # python datetime.datetime.timetuple()
+    # Week is MTWTFSS, 
+    # Monday = 0 (for python datetime.datetime.timetuple())
     dayOfWeekExpected = [0,0,0,1,0,0,0] #Thu
```
Comment (breznak) [inline py/tests/encoders/date_test.py]: Thursday is now again back to being 4th day of week in the representation, which is nice.

File: py/htm/encoders/date.py (hunk 7 above)
Comment (ctrl-z-9000-times) [reply py/htm/encoders/date.py]: This way works even if the dimensions are different types, such as tuple vs list.

This class keeps its dimensions as a tuple, the SDR class keeps its dimensions as a list (b/c of pybind11).

File: bindings/py/tests/encoders/scalar_encoder_test.py (hunk 6 above)
Comment (ctrl-z-9000-times) [reply bindings/py/tests/encoders/scalar_encoder_test.py]: I don't want the users to need to import the bindings.  Instead I want everything to appear as a single python package, with the bindings as a hidden/unused backend.  I think it would be confusing for the users to see imports from both `htm.encoders` and `htm.bindings.encoders` because then they won't know where to look to find things.

File: py/htm/encoders/date.py (hunk 9 above)
Comment (ctrl-z-9000-times) [reply py/htm/encoders/date.py]: `Round towards noon, not midnight, this means similarity of representations changes at midnights, not noon.`

Actually, the similarity changes every hour throughout the day.  The effect you're describing here only happens in the unrealistic case of having a single bit per day.

Comment (breznak) [reply py/htm/encoders/date.py]: ok, true. "In the too-few-bits scenario, when the 2 representations don't overlap"

**Comment by breznak** [comment]
Thank you for the quick reviews, David!
